### ISSUE TITLE
Order of the code completion menu not correct

### ISSUE THREAD

**Issue by m1ga** [pulsar-edit/pulsar#994]
### Thanks in advance for your bug report!

- [X] Have you reproduced issue in [safe mode](https://pulsar-edit.dev/docs/atom-archive/faq/sections/what-does-safe-mode-do.html)?
- [X] Have you used the [debugging guide](https://pulsar-edit.dev/docs/atom-archive/hacking-atom/#debugging) to try to resolve the issue?
- [X] Have you checked our [FAQs](https://pulsar-edit.dev/docs/launch-manual/sections/faq/) to make sure your question isn't answered there?
- [X] Have you checked to make sure your issue does not already [exist](https://github.com/issues?q=is%3Aissue+user%3Apulsar-edit+is%3Aopen)?
- [X] Have you checked you are on the latest release of Pulsar?

### What happened?

The order of the code completion menu is not in the best/correct order. Here I'm using the Titanium SDK package to complete a JS method. I want to use `createWindow` and when I type `Ti.UI.createW` the correct one with a capital W is only the 2nd suggestion. It ranks `createView` higher.
Is there a way to change that order? Or do I need to change something in the package itself?

![Screenshot_20240429_204622](https://github.com/pulsar-edit/pulsar/assets/4334997/2608c2bb-e120-4be7-ad97-d6a37451b4e6)

```
Pulsar  : 1.116.0
Electron: 12.2.3
Chrome  : 89.0.4389.128
Node    : 14.16.0
```

### Pulsar version

1.116.0

### Which OS does this happen on?

🐧 Red Hat based (Fedora, Alma, RockyLinux, CentOS Stream, etc.)

### OS details

Linux fedora 6.8.7-300.fc40.x86_64

### Which CPU architecture are you running this on?

x86_64/AMD64

### What steps are needed to reproduce this?

* install [the](https://web.pulsar-edit.dev/packages/titanium)
* start a new JS file
* start typing `Ti.UI.createW`

**Comment by savetheclocktower** [comment]
Sounds like this might be another difference in fuzzy matching between `atom.ui.fuzzyMatcher` and `fuzzaldrin`. @mauricioszabo, any ideas?

**Comment by mauricioszabo** [comment]
This is actually not on our side. Autocomplete _can_ sort, but packages can override it.

This is the code at TitaniumSDK that does the sorting: https://github.com/tidev/pulsar-titanium/blob/master/lib/providers/autoCompleteHelper.js#L20-L29

As you can see, it first sorts by priority, then by word length. If you want, you can ask for the author to change this sorting algorithm, or upvote the suggestion I did on author's package: https://github.com/tidev/pulsar-titanium/issues/684

**Comment by savetheclocktower** [comment]
> This is actually not on our side. Autocomplete _can_ sort, but packages can override it.
 
I thought so, too, but all their providers have `filterSuggestions: true`, so now I'm not sure.

**Comment by mauricioszabo** [comment]
So, I tried this query on our internal API, and this is how it sorted:

![image](https://github.com/pulsar-edit/pulsar/assets/138037/82a414bd-7d82-4e10-9db3-b8f693abf60a)

The score of `createWindow` is higher on both cases. I'm not sure how it would be sorted in a different way... I'm also not sure what Titanium is, otherwise I could try to reproduce on a dev build and see what's happening

**Comment by m1ga** [comment]
you should be able to create a Titanium project with:
```js
sudo npm i -g titanium
ti sdk install latest
ti create -t app  -n test --id test.test -p android -u -d .
```
it should create a test folder which you can open with Pulsar and the package. Then inside the `Resources` folder is a app.js file you can edit.

**Comment by m1ga** [comment]
Is it perhaps a Linux issue?

https://github.com/pulsar-edit/pulsar/assets/4334997/f3a28225-5291-4657-b515-941799ffbc93

I have this with CSS/SASS too. `currencolor` won't go away

**Comment by savetheclocktower** [comment]
> I have this with CSS/SASS too. `currencolor` won't go away

I don't think this is coming from `autocomplete-css` because I can't get it to suggest `currentcolor` (it wants to suggest the more accurate `currentColor`, and it doesn't make it the first option). I thought that you might have the built-in provider enabled (the generic one that comes with `autocomplete-plus` that just suggests words that are already present in the file), but if it were coming from that provider I don't think it'd have the “V” to the left.

I suspect the suggestion is coming from the `titanium` package. I can't see an obvious place that it would come from, but I do see that `titanium` is doing a strange thing. It tries to offer completions for some Titanium-specific things, but if those things aren't present at the cursor, it instead tries to suggest things that a CSS completion provider would suggest. That feels strange because it should just be offering its own suggestions — not trying to take the place of something like `autocomplete-scss`.

@m1ga, please do me a quick favor, as silly as it feels to ask for: please temporarily disable `titanium` and confirm that this issue goes away. If there's no `currentcolor` suggestion while `titanium` is disabled, we can be certain that package is the culprit.

Assuming it's the culprit, we have some options, since I notice that [you're a contributor to this package](https://github.com/tidev/pulsar-titanium).

If it were me, [I'd look at this behavior](https://github.com/tidev/pulsar-titanium/blob/master/lib/providers/styleAutoCompleteProvider.js#L47-L49) — that's where it decides to supply CSS value completion suggestions. It might be worth making that a configuration option so that you have the ability to tell it not to try to offer CSS value completions. In general, you'll get better results with a dedicated CSS autocompletion provider. (The builtin `autocomplete-css` doesn't do so well with the new Tree-sitter grammars, but I just published [pulsar-ide-css-alpha](https://web.pulsar-edit.dev/packages/pulsar-ide-css-alpha) as an experiment last week, and it's equally good at offering completions regardless of grammar.)

At any rate, if you comment out those lines for testing, I suspect that will stop it from suggesting `currentColor`.

And if you want to help us figure out whether the weird ordering is `titanium`’s fault or ours: I'd suggest _temporarily_ turning [this `sort` method](https://github.com/tidev/pulsar-titanium/blob/master/lib/providers/autoCompleteHelper.js#L21) into one that simply returns its input without changing it. If that results in a more pleasing order, you can remove all of `titanium`’s attempts to sort completions and let our fuzzy-matching do all the work. If the order is the same, then that'd go a long way toward proving that this is our problem and not the package's.

**Comment by m1ga** [comment]
Thank you for the detailed reply. I've disabled/completly remove Titanium but it still shows `currentcolor`. I don't mind it beeing there but it shouldn't be the first suggestion if the other suggestions are better. E.g. if I type `wi` to write `width` it shows up like this:

![Bildschirmfoto_20240626_211109](https://github.com/pulsar-edit/pulsar/assets/4334997/deccbcf8-8f32-43cd-8de4-54b9068503a2)

even if `width` and `windows` are better suggestions.
My package list:
* atom-beautify
* atom-ide-datatip
* atomic-management
* file-icons
* highlight-selected
* hyperclick
* itentions
* js-hyperclick
* language-nunjucks
* language-vue
* linter
* minimap
* minimap-git-diff
* minimap-highlight-selected
* pigments
* smart-html-tags
* sort-lines
* sync-settings

**Comment by savetheclocktower** [comment]
That's a fair one; I can reproduce that. in that instance, I think it's favoring a mediocre result from one provider (`autocomplete-snippets`) over a much better one from another provider (`autocomplete-css`). I'm going to dig into this a bit.

This wouldn't necessarily have the same root cause as the initial bug you filed — in which the results _within a single provider_ were ordered unintuitively. But we'll see!

**Comment by savetheclocktower** [comment]
Yeah, the deal in this case is that `autocomplete-plus` will show all the results from one provider above all the results from the next provider. It will not interleave results from providers, nor will it elevate the provider with the best result to the top, even in cases where the two providers have the same priority.

I think interleaving results should be an option (if not the default behavior). After about ten minutes of tinkering, I was able to produce this outcome:

<img width="353" alt="Screenshot 2024-06-26 at 2 39 51 PM" src="https://github.com/pulsar-edit/pulsar/assets/3450/1a16c25e-fec2-4a83-bff4-234357f02afc">

I'm going to file a ticket for this. It might be time for `autocomplete-plus-plus`.

---

I'm still dying to know which provider is unhelpfully surfacing `currentcolor`. @m1ga, try going into `autocomplete-plus` settings and making sure “Enable Built-In Provider“ is unchecked just so I can rule that out.

**Comment by mauricioszabo** [comment]
And this is because the plug-in is called `autocomplete-plus`. Imagine how weird was `autocomplete` without the `-plus`, right? :laughing: 

Anyway... I feel we need to fix this in `autocomplete-plus` indeed. What is **very weird** is that `autocomplete-plus` don't require a `score` field for completions, so it might not be clear how to "sort" these results... as soon as you open the issue, @savetheclocktower, let's discuss how to best approach this case. I might have some ideas :)

**Comment by savetheclocktower** [comment]
Opened #1037.

**Comment by m1ga** [comment]
> I'm still dying to know which provider is unhelpfully surfacing currentcolor. @m1ga, try going into autocomplete-plus settings and making sure “Enable Built-In Provider“ is unchecked just so I can rule that out.

Unchecked it but it still show `currentcolor`. BTW: it links to https://developer.mozilla.org/en-US/docs/Web/CSS/color#values
